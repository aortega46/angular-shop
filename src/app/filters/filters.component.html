<section class="flex justify-between">
  <div class="flex items-center gap-1">
    <label for="minPrice">Price from: </label>
    <input
      #minPrice
      name="minPrice"
      type="range"
      min="0"
      max="2000"
      value="{{ (productsService.filters | async)!.minPrice }}"
      (input)="handleChangeMinPrice($event)" />
    <span>${{ (productsService.filters | async)!.minPrice }}</span>
  </div>

  <div class="flex items-center gap-1">
    <label for="category">Category</label>
    <select
      #category
      class="text-black"
      name="category"
      (input)="handleChangeCategory($event)">
      <option value="all">All</option>
      <option value="smartphones">Smartphones</option>
      <option value="laptops">Laptops</option>
    </select>
  </div>
</section>
