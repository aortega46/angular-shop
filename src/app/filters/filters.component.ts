import { Component, inject } from '@angular/core'
import { ProductsService } from '../services/products.service'
import { AsyncPipe } from '@angular/common'

@Component({
  selector: 'app-filters',
  standalone: true,
  imports: [AsyncPipe],
  templateUrl: './filters.component.html',
})
export class FiltersComponent {
  productsService = inject(ProductsService)

  handleChangeMinPrice(event: Event) {
    const minPrice = +(event.target as HTMLInputElement).value
    this.productsService.updateFilters({ minPrice })
  }

  handleChangeCategory(event: Event) {
    const category = (event.target as HTMLInputElement).value
    this.productsService.updateFilters({ category })
  }
}
