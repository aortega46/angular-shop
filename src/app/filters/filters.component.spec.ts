import { ComponentFixture, TestBed } from '@angular/core/testing'

import { FiltersComponent } from './filters.component'
import { By } from '@angular/platform-browser'
import { ProductsService } from '../services/products/products.service'

describe('FiltersComponent', () => {
  let component: FiltersComponent
  let fixture: ComponentFixture<FiltersComponent>
  let compiled: HTMLElement

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [FiltersComponent],
      providers: [ProductsService],
    }).compileComponents()

    fixture = TestBed.createComponent(FiltersComponent)
    component = fixture.componentInstance
    fixture.detectChanges()

    compiled = fixture.nativeElement
  })

  it('should create', () => {
    expect(component).toBeTruthy()
  })

  it('should match snapshot', () => {
    expect(compiled).toMatchSnapshot()
  })

  it('should call handleChangeMinPrice when user updates minPrice filter', () => {
    const spyHandleChangeMinPrice = jest.spyOn(
      component,
      'handleChangeMinPrice'
    )

    const mockProductsService = TestBed.inject(ProductsService)
    const spyUpdateFilters = jest.spyOn(mockProductsService, 'updateFilters')

    const inputElement = fixture.debugElement.query(By.css('input'))
    inputElement.triggerEventHandler('input', { target: { value: 500 } })

    expect(spyHandleChangeMinPrice).toHaveBeenCalled()
    expect(spyUpdateFilters).toHaveBeenCalledWith({ minPrice: 500 })
  })

  it('should call handleChangeCategory when user updates category filter', () => {
    const spyHandleChangeMinPrice = jest.spyOn(
      component,
      'handleChangeCategory'
    )

    const mockProductsService = TestBed.inject(ProductsService)
    const spyUpdateFilters = jest.spyOn(mockProductsService, 'updateFilters')

    const inputElement = fixture.debugElement.query(By.css('select'))
    inputElement.triggerEventHandler('input', { target: { value: 'laptops' } })

    expect(spyHandleChangeMinPrice).toHaveBeenCalled()
    expect(spyUpdateFilters).toHaveBeenCalledWith({ category: 'laptops' })
  })
})
