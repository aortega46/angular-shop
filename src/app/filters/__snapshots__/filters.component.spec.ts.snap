// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FiltersComponent should match snapshot 1`] = `
<div
  id="root1"
>
  <section
    class="flex flex-col justify-between gap-2 sm:flex-row"
  >
    <div
      class="flex items-center gap-2"
    >
      <label
        for="minPrice"
      >
        Price from: 
      </label>
      <input
        max="2000"
        min="0"
        name="minPrice"
        type="range"
      />
      <span>
        $0
      </span>
    </div>
    <div
      class="flex items-center gap-2"
    >
      <label
        for="category"
      >
        Category
      </label>
      <select
        class="text-black"
        name="category"
      >
        <option
          value="all"
        >
          All
        </option>
        <option
          value="smartphones"
        >
          Smartphones
        </option>
        <option
          value="laptops"
        >
          Laptops
        </option>
      </select>
    </div>
  </section>
</div>
`;
